components/browser/session/src/main/java/mozilla/components/browser/session/SessionObserver.kt

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.session
+
+/**
+ * This class is a combination of [Session.Observer] and
+ * [SessionManager.Observer]. It provides functionality to observe changes to a
+ * specified or selected session, and automatically takes care of switching
+ * over the observer in case a different session is selected (see [observe] and
+ * [observeSelected]).
+ *
+ * @property activeSession the currently observed session
+ * @property sessionManager the application's session manager
+ */
+abstract class SessionObserver(
```

Comment:
nit: I wonder if we can find a more explicit name. The difference between `Session.Observer` and `SessionObserver` is very subtle. And you can actually accidentally use `SessionObserver` instead of `Session.Observer` because it implements that. :)

Not sure what a good name would be though. `SingleSessionObserver`, `FollowingSessionObserver`, `AutoSessionObserver`, `SelectedSessionObserver` (it can do more than that though).